src/utils/mongoWatcher.ts

```diff
@@ -0,0 +1,141 @@
+   * @param {RabbitDataType}  rabbitData  - rabbit data
+   * @param {MTROptions}      options     - mongo to rabbit options
+   */
+  constructor(mongoData: MongoDataType, rabbitData: RabbitDataType, options: MTROptions) {
+    this.mongoData = mongoData;
+    this.rabbitData = rabbitData;
+    this.options = options;
+    if (mongoData.healthCheckInterval) this.healthCheckInterval = mongoData.healthCheckInterval;
+  }
+
+  /**
+   * Starts mongo connection
+   */
+  async mongoConnection() {
+    while (true) {
+      try {
+        logger.log(
+          `try connect mongo collection: ${this.mongoData.collectionName}, uri: ${this.mongoData.connectionString}`
+        );
+        await mongoose.connect(this.mongoData.connectionString, mongoOptions);
+        logger.log(`successful connection to mongo on connectionString: ${this.mongoData.connectionString}`);
+
+        this.initWatch();
+        break;
+      } catch (error) {
+        console.log(`can't connect to mongo. Retrying in ${this.healthCheckInterval}ms`);
+        console.log(error);
+
+        await sleep(this.healthCheckInterval);
+      }
+    }
+  }
+
+  /**
+   * Initiate collection watcher change stream from last event id. if none found, returns undefiend
+   * @param {String} collectionName - collection name that the mongo watches
+   */
+  async initiateChangeStreamStartTime(collectionName: string) {
+    // Get the last event
+    const latestEvent: any = await changeStreamTrackerModel(collectionName).findOne({}).sort({ createdAt: -1 });
+    const latestEventId = latestEvent && latestEvent.eventId;
+
+    return latestEventId;
+  }
+
+  /**
+   * Initializes the mongo watcher, given the mongo data.
+   */
+  async initWatch() {
+    // Get the last event id that successfully sent to rabbit
+    const lastEventId = await this.initiateChangeStreamStartTime(this.mongoData.collectionName);
+
+    // Select DB and Collection
+    const connectionObject: IConnectionStringParameters = csParser.parse(this.mongoData.connectionString);
+    const collection = collectionModel(this.mongoData.collectionName);
+
+    // Define collection change stream settings
+    const pipeline = [{ $match: { 'ns.db': connectionObject.endpoint, 'ns.coll': this.mongoData.collectionName } }];
+    const optionsStream: ChangeStreamOptions = { fullDocument: 'updateLookup' };
+
+    if (lastEventId) {
+      const startAfterToken: ResumeToken = {};
+      (startAfterToken as any)['_data'] = lastEventId;
+      optionsStream.startAfter = startAfterToken;
+    }
+
+    const changeStream = collection.watch(pipeline, optionsStream);
+
+    // start listen to changes
+    changeStream
+      .on('change', async (event: ChangeEvent<any>) => {
+        logger.log(`got mongo event: ${event.operationType}, at collection:${this.mongoData.collectionName}`);
+
+        try {
+          // Try send msg to all queues
+          await Promise.all(
```

Comment:
Promise.allSettled()

Reply:
In this case it iisn't necessaryto use "allSettled" because we would like to resend the same message if one of the missions to queues fails.

Reply:
But it will `throw` the error and won't continue the `pending` `Promise` status which will cancel the next `rejected` and `resloved` promises